File: stdlib/ksc/ksc.py
# test644 : stdlib5.ksc

import zlib
import os

class toolbox:
    def __init__(self):
        self.prehead = webpbase() # prehead + pad 512nB
        self.common = b"KSC5" # common sign 4B
        self.subtype = b"\x00\x00\x00\x00" # subtype sign 4B
        self.reserved = b"\x00\x00\x00\x00\x00\x00\x00\x00" # reserved 8B

        self.headp = 512 # mainheader start point (512n)
        self.rsize = 0 # valid data size
        self.path = "" # file path

        self.predetect = False # read chunk info ahead
        self.chunkpos = [ ] # chunk (8B + nB) pos
        self.chunksize = [ ] # chunk data (nB) size

    # read by file path
    def readf(self):
        if self.path == "":
            raise Exception("NoSuchFile")
        self.prehead, self.headp, size = b"", 0, os.path.getsize(self.path)

        with open(self.path, "rb") as f:
            while size >= self.headp + 4:
                temp = f.read(4)
                if temp == self.common:
                    break
                else:
                    self.prehead = self.prehead + temp + f.read(508)
                    self.headp = self.headp + 512
            if size < self.headp + 16:
                raise Exception("InvalidKSC5")
            
            self.subtype = f.read(4)
            self.reserved = f.read(8)
            self.rsize = size - self.headp

            self.chunkpos, self.chunksize, pos = [ ], [ ], self.headp + 16
            if self.predetect:
                while size >= pos + 8:
                    temp = f.read(8)
                    if temp == b"\xff\xff\xff\xff\xff\xff\xff\xff":
                        break
                    else:
                        temp = decode(temp)
                        self.chunkpos.append(pos)
                        self.chunksize.append(temp)
                        pos = pos + 8 + temp
                        f.seek(pos)

    # init & write mainhead by file path
    def writef(self):
        self.headp = len(self.prehead)
        self.rsize = 0
        self.chunkpos, self.chunksize = [ ], [ ]
        if (self.headp % 512 != 0) or (len(self.common) != 4) or (len(self.subtype) != 4) or (len(self.reserved) != 8):
            raise Exception("InvalidKSC5")
        if self.path == "":
            raise Exception("NoSuchFile")
        
        with open(self.path, "wb") as f:
            f.write(self.prehead + self.common + self.subtype + self.reserved)

    # write one chunk by content of file, empty path : end (8x FF)
    def addf(self, path):
        if path == "":
            with open(self.path, "ab") as f:
                f.write(b"\xff\xff\xff\xff\xff\xff\xff\xff")
        else:
            size = os.path.getsize(path)
            with open(self.path, "ab") as f:
                f.write( encode(size, 8) )
                with open(path, "rb") as t:
                    for i in range(0, size // 10485760):
                        f.write( t.read(10485760) )
                    f.write( t.read(size % 10485760) )

    # write one chunk by binary data
    def linkf(self, data):
        with open(self.path, "ab") as f:
            f.write( encode(len(data), 8) )
            f.write(data)

    # read by binary data
    def readb(self, data):
        self.prehead, self.headp, size = b"", 0, len(data)

        while size >= self.headp + 4:
            if data[self.headp:self.headp + 4] == self.common:
                break
            else:
                self.prehead = self.prehead + data[self.headp:self.headp + 512]
                self.headp = self.headp + 512
        if size < self.headp + 16:
            raise Exception("InvalidKSC5")
            
        self.subtype = data[self.headp + 4:self.headp + 8]
        self.reserved = data[self.headp + 8:self.headp + 16]
        self.rsize = size - self.headp

        self.chunkpos, self.chunksize, pos = [ ], [ ], self.headp + 16
        if self.predetect:
            while size >= pos + 8:
                temp = data[pos:pos + 8]
                if temp == b"\xff\xff\xff\xff\xff\xff\xff\xff":
                    break
                else:
                    temp = decode(temp)
                    self.chunkpos.append(pos)
                    self.chunksize.append(temp)
                    pos = pos + 8 + temp

    # init & write mainhead by binary data
    def writeb(self):
        self.headp = len(self.prehead)
        self.rsize = 0
        self.chunkpos, self.chunksize = [ ], [ ]
        if (self.headp % 512 != 0) or (len(self.common) != 4) or (len(self.subtype) != 4) or (len(self.reserved) != 8):
            raise Exception("InvalidKSC5")
        
        return b"".join( [self.prehead, self.common, self.subtype, self.reserved] )

    # write one chunk by content of file, empty path : end (8x FF)
    def addb(self, stream, path):
        if path == "":
            return stream + b"\xff\xff\xff\xff\xff\xff\xff\xff"
        else:
            size = os.path.getsize(path)
            with open(path, "rb") as f:
                temp = f.read()
            return b"".join( [stream, encode(size, 8), temp] )

    # write one chunk by binary data
    def linkb(self, stream, data):
        return b"".join( [stream, encode(len(data), 8), data] )

# little endian encoding, I -> B
def encode(num, length):
    temp = [0] * length
    for i in range(0, length):
        temp[i] = num % 256
        num = num // 256
    return bytes(temp)

# little endian decoding, B -> I
def decode(data):
    temp = 0
    for i in range( 0, len(data) ):
        if data[i] != 0:
            temp = temp + data[i] * 256 ** i
    return temp

# crc32
def crc32hash(data):
    return encode(zlib.crc32(data), 4)

# basic KSC5 prehead webp data
def webpbase():
    data0 = b""
    data0 = data0 + b"\x52\x49\x46\x46\xbc\x01\x00\x00\x57\x45\x42\x50\x56\x50\x38\x20\xb0\x01\x00\x00\x50\x09\x00\x9d\x01\x2a\x40\x00\x40\x00\x3e\x69\x2c\x90\x45\xa4\x22\xa1\x9a\xfa"
    data0 = data0 + b"\x34\xcc\x40\x06\x84\xb3\x80\x67\x2c\xd1\xff\xfa\x7a\x71\x96\xf7\x28\x36\x82\x88\x0e\x47\xbf\x98\xba\x75\x08\x2e\xed\xcf\x6a\xfd\x0d\xf3\xed\x21\xc2\x2e\xfc\xc0"
    data0 = data0 + b"\xc4\x87\x23\xf1\xf7\x5b\x9a\x84\xe6\x14\x3f\x5b\xc1\x1b\x75\x50\xb9\xae\xe4\x94\xb6\x49\xc0\x00\xec\x8f\xc2\xb9\x12\xdc\x9c\x2a\x54\x4d\x1f\xdb\xe8\xf0\xdd\x00"
    data0 = data0 + b"\xc3\xca\xff\x48\xc4\x4b\xf9\x50\x76\x84\x72\x68\xdf\xd4\xed\xc9\x9c\xe0\x78\xf5\xf4\x61\x6e\x63\x52\xd8\x15\x7c\xe5\x23\xd5\x3b\x97\x09\x67\x89\x6a\x8b\xe3\x18"
    data0 = data0 + b"\xb0\x8a\xa4\x41\x57\x8a\x97\x6b\xc8\x06\x2c\xd0\xd6\xce\xe6\xbb\x42\x6f\x8c\x01\x5f\xa5\x6f\x34\xcf\xbe\x81\x92\x33\x3f\x9c\xaa\x25\x16\x29\x53\x27\xa1\x48\xd8"
    data0 = data0 + b"\x8e\x58\xc3\xfd\x01\x34\x52\x36\x6c\x25\xfc\xc0\x68\xed\x2c\x0d\x95\x9a\x8a\xd3\x89\x90\xf7\x2a\xa2\x3b\xb0\xc6\xd4\xf5\x19\x51\x34\xbc\x72\x2f\x2d\xb8\x46\x32"
    data0 = data0 + b"\x8b\x59\x33\x08\xb1\x76\x37\x77\x63\x41\xa5\x0e\xee\x55\x7d\x73\x2e\x5b\x8e\x50\x99\x41\xfe\x96\xf2\x34\x36\x3a\xda\xd0\xd9\x17\x89\x13\xa0\x9b\xab\xf9\x48\xbd"
    data0 = data0 + b"\xca\xf8\x36\x33\x0e\x98\xd9\xe8\x95\x94\x2a\xb0\x92\x42\x01\x47\xe0\x53\xa7\xfd\x65\x38\x6a\x65\x6b\x39\x3c\x6c\xca\x2d\x51\x32\x33\xcd\x68\x5c\xdf\x23\xae\x58"
    data0 = data0 + b"\x4f\x0e\x7d\xbb\xdb\x98\x56\x62\xc4\x61\x94\x2a\xdd\x51\xcd\x79\xeb\x38\x99\x4a\x4a\x6b\xa6\x2d\x7b\xc5\xba\xbb\x92\x9a\xf4\x7d\xb7\xdc\x57\xc6\xfd\xe9\x66\x2a"
    data0 = data0 + b"\x8f\xf8\x04\xf3\x28\x33\x0c\xb5\x12\x8b\x39\xd0\xaa\xce\xae\xac\x90\x77\x2d\xe3\xf2\xb0\x23\x21\xa7\x4f\x26\xfb\x51\x45\x9e\x27\xc3\xb0\x6a\x6a\x0b\xd3\xfd\x3c"
    data0 = data0 + b"\x31\x65\x05\xfa\x21\x64\xa0\x36\x1e\x5d\x8b\xd4\xe9\xd7\xe1\x59\x72\x2f\x0c\x1c\x47\x3e\x1c\xbf\xe0\x15\xc0\x57\xa2\x30\x47\xc6\x4d\x02\xbd\xbf\x6f\xb6\xfc\x12"
    data0 = data0 + b"\x10\x4c\x3b\xde\xea\xdd\x7b\x00\x00\x00\x00\x00"

    data0 = data0 + b"\x00" * ( 512 - len(data0) )
    return data0

File: stdlib/ksc/ksc.go
// test645 : stdlib5.ksc

package ksc

import (
	"bytes"
	"errors"
	"hash/crc32"
	"stdlib5/kio"
)

type toolbox struct {
	Prehead  []byte // prehead + pad 512nB
	Common   []byte // common sign 4B
	Subtype  []byte // subtype sign 4B
	Reserved []byte // reserved 8B

	Headp int    // mainheader start point (512n)
	Rsize int    // valid data size
	Path  string // file path

	Predetect bool  // read chunk info ahead
	Chunkpos  []int // chunk (8B + nB) pos
	Chunksize []int // chunk data (nB) size
}

// read by file path
func (tbox *toolbox) Readf() error {
	var temp []byte
	tbox.Prehead = make([]byte, 0)
	tbox.Headp = 0
	size := kio.Size(tbox.Path)
	if (tbox.Path == "") || (size == -1) {
		return errors.New("NoSuchFile")
	}

	f, err := kio.Open(tbox.Path, "r")
	if err == nil {
		defer f.Close()
	} else {
		return err
	}
	for size >= tbox.Headp+4 {
		temp, _ = kio.Read(f, 4)
		if kio.Bequal(temp, tbox.Common) {
			break
		} else {
			tbox.Prehead = append(tbox.Prehead, temp...)
			temp, _ = kio.Read(f, 508)
			tbox.Prehead = append(tbox.Prehead, temp...)
			tbox.Headp = tbox.Headp + 512
		}
	}
	if size < tbox.Headp+16 {
		return errors.New("InvalidKSC5")
	}

	tbox.Subtype, _ = kio.Read(f, 4)
	tbox.Reserved, _ = kio.Read(f, 8)
	tbox.Rsize = size - tbox.Headp

	tbox.Chunkpos = make([]int, 0)
	tbox.Chunksize = make([]int, 0)
	pos := tbox.Headp + 16
	if tbox.Predetect {
		for size >= pos+8 {
			temp, _ = kio.Read(f, 8)
			if kio.Bequal(temp, []byte{255, 255, 255, 255, 255, 255, 255, 255}) {
				break
			} else {
				ti := Decode(temp)
				tbox.Chunkpos = append(tbox.Chunkpos, pos)
				tbox.Chunksize = append(tbox.Chunksize, ti)
				pos = pos + 8 + ti
				te0, te1 := f.Seek(int64(pos), 0)
				if (int(te0) != pos) || (te1 != nil) {
					return te1
				}
			}
		}
	}
	return nil
}

// init & write mainhead by file path
func (tbox *toolbox) Writef() error {
	tbox.Headp = len(tbox.Prehead)
	tbox.Rsize = 0
	tbox.Chunkpos = make([]int, 0)
	tbox.Chunksize = make([]int, 0)
	if (tbox.Headp%512 != 0) || (len(tbox.Common) != 4) || (len(tbox.Subtype) != 4) || (len(tbox.Reserved) != 8) {
		return errors.New("InvalidKSC5")
	}
	if tbox.Path == "" {
		return errors.New("NoSuchFile")
	}

	f, err := kio.Open(tbox.Path, "w")
	if err == nil {
		defer f.Close()
	} else {
		return err
	}
	temp := bytes.Join([][]byte{tbox.Prehead, tbox.Common, tbox.Subtype, tbox.Reserved}, make([]byte, 0))
	_, err = kio.Write(f, temp)
	return err
}

// write one chunk by content of file, empty path : end (8x FF)
func (tbox *toolbox) Addf(path string) error {
	f, err := kio.Open(tbox.Path, "a")
	if err == nil {
		defer f.Close()
	} else {
		return err
	}

	if path == "" {
		_, err = kio.Write(f, []byte{255, 255, 255, 255, 255, 255, 255, 255})
		return err
	} else {
		t, err := kio.Open(path, "r")
		if err == nil {
			defer t.Close()
		} else {
			return err
		}

		size := kio.Size(path)
		_, err = kio.Write(f, Encode(size, 8))
		if err != nil {
			return err
		}

		for i := 0; i < size/10485760; i++ {
			temp, err := kio.Read(t, 10485760)
			if err != nil {
				return err
			}
			_, err = kio.Write(f, temp)
			if err != nil {
				return err
			}
		}
		temp, err := kio.Read(t, size%10485760)
		if err != nil {
			return err
		}
		_, err = kio.Write(f, temp)
		return err
	}
}

// write one chunk by binary data
func (tbox *toolbox) Linkf(data []byte) error {
	f, err := kio.Open(tbox.Path, "a")
	if err == nil {
		defer f.Close()
	} else {
		return err
	}

	_, err = kio.Write(f, Encode(len(data), 8))
	if err != nil {
		return err
	}
	_, err = kio.Write(f, data)
	return err
}

// read by binary data
func (tbox *toolbox) Readb(data []byte) error {
	var temp []byte
	tbox.Prehead = make([]byte, 0)
	tbox.Headp = 0
	size := len(data)

	for size >= tbox.Headp+4 {
		temp = data[tbox.Headp : tbox.Headp+4]
		if kio.Bequal(temp, tbox.Common) {
			break
		} else {
			tbox.Prehead = append(tbox.Prehead, data[tbox.Headp:tbox.Headp+512]...)
			tbox.Headp = tbox.Headp + 512
		}
	}
	if size < tbox.Headp+16 {
		return errors.New("InvalidKSC5")
	}

	tbox.Subtype = make([]byte, 4)
	copy(tbox.Subtype, data[tbox.Headp+4:tbox.Headp+8])
	tbox.Reserved = make([]byte, 8)
	copy(tbox.Reserved, data[tbox.Headp+8:tbox.Headp+16])
	tbox.Rsize = size - tbox.Headp

	tbox.Chunkpos = make([]int, 0)
	tbox.Chunksize = make([]int, 0)
	pos := tbox.Headp + 16

	if tbox.Predetect {
		for size >= pos+8 {
			temp = data[pos : pos+8]
			if kio.Bequal(temp, []byte{255, 255, 255, 255, 255, 255, 255, 255}) {
				break
			} else {
				ti := Decode(temp)
				tbox.Chunkpos = append(tbox.Chunkpos, pos)
				tbox.Chunksize = append(tbox.Chunksize, ti)
				pos = pos + 8 + ti
			}
		}
	}
	return nil
}

// init & write mainhead by binary data
func (tbox *toolbox) Writeb() ([]byte, error) {
	tbox.Headp = len(tbox.Prehead)
	tbox.Rsize = 0
	tbox.Chunkpos = make([]int, 0)
	tbox.Chunksize = make([]int, 0)
	if (tbox.Headp%512 != 0) || (len(tbox.Common) != 4) || (len(tbox.Subtype) != 4) || (len(tbox.Reserved) != 8) {
		return nil, errors.New("InvalidKSC5")
	}
	return bytes.Join([][]byte{tbox.Prehead, tbox.Common, tbox.Subtype, tbox.Reserved}, make([]byte, 0)), nil
}

// write one chunk by content of file, empty path : end (8x FF)
func (tbox *toolbox) Addb(stream []byte, path string) ([]byte, error) {
	var out []byte
	if path == "" {
		out = []byte{255, 255, 255, 255, 255, 255, 255, 255}
	} else {
		size := kio.Size(path)
		out = Encode(size, 8)
		t, err := kio.Open(path, "r")
		if err == nil {
			defer t.Close()
		} else {
			return nil, err
		}

		t0, t1 := kio.Read(t, -1)
		if t1 != nil {
			return nil, t1
		}
		out = append(out, t0...)
	}
	return append(stream, out...), nil
}

// write one chunk by binary data
func (tbox *toolbox) Linkb(stream []byte, data []byte) []byte {
	return bytes.Join([][]byte{stream, Encode(len(data), 8), data}, make([]byte, 0))
}

// Init and return toolbox
func Initksc() toolbox {
	var out toolbox
	out.Prehead = Webpbase()
	out.Common = []byte("KSC5")
	out.Subtype = make([]byte, 4)
	out.Reserved = make([]byte, 8)
	out.Headp = 512
	out.Rsize = 0
	out.Path = ""
	out.Predetect = false
	out.Chunkpos = make([]int, 0)
	out.Chunksize = make([]int, 0)
	return out
}

// little endian encoding, I -> B
func Encode(num int, length int) []byte {
	temp := make([]byte, length)
	for i := 0; i < length; i++ {
		temp[i] = byte(num % 256)
		num = num / 256
	}
	return temp
}

// little endian decoding, B -> I
func Decode(data []byte) int {
	temp := 0
	for i, r := range data {
		if r != 0 {
			exp := 1
			for j := 0; j < i; j++ {
				exp = exp * 256
			}
			temp = temp + int(r)*exp
		}
	}
	return temp
}

// crc32
func Crc32hash(data []byte) []byte {
	temp := int(crc32.ChecksumIEEE(data))
	return Encode(temp, 4)
}

// basic KSC5 prehead webp data
func Webpbase() []byte {
	var temp []byte
	temp = append(temp, 82, 73, 70, 70, 188, 1, 0, 0, 87, 69, 66, 80, 86, 80, 56, 32, 176, 1, 0, 0, 80, 9, 0, 157, 1, 42, 64, 0, 64, 0, 62, 105, 44, 144, 69, 164, 34, 161, 154, 250)
	temp = append(temp, 52, 204, 64, 6, 132, 179, 128, 103, 44, 209, 255, 250, 122, 113, 150, 247, 40, 54, 130, 136, 14, 71, 191, 152, 186, 117, 8, 46, 237, 207, 106, 253, 13, 243, 237, 33, 194, 46, 252, 192)
	temp = append(temp, 196, 135, 35, 241, 247, 91, 154, 132, 230, 20, 63, 91, 193, 27, 117, 80, 185, 174, 228, 148, 182, 73, 192, 0, 236, 143, 194, 185, 18, 220, 156, 42, 84, 77, 31, 219, 232, 240, 221, 0)
	temp = append(temp, 195, 202, 255, 72, 196, 75, 249, 80, 118, 132, 114, 104, 223, 212, 237, 201, 156, 224, 120, 245, 244, 97, 110, 99, 82, 216, 21, 124, 229, 35, 213, 59, 151, 9, 103, 137, 106, 139, 227, 24)
	temp = append(temp, 176, 138, 164, 65, 87, 138, 151, 107, 200, 6, 44, 208, 214, 206, 230, 187, 66, 111, 140, 1, 95, 165, 111, 52, 207, 190, 129, 146, 51, 63, 156, 170, 37, 22, 41, 83, 39, 161, 72, 216)
	temp = append(temp, 142, 88, 195, 253, 1, 52, 82, 54, 108, 37, 252, 192, 104, 237, 44, 13, 149, 154, 138, 211, 137, 144, 247, 42, 162, 59, 176, 198, 212, 245, 25, 81, 52, 188, 114, 47, 45, 184, 70, 50)
	temp = append(temp, 139, 89, 51, 8, 177, 118, 55, 119, 99, 65, 165, 14, 238, 85, 125, 115, 46, 91, 142, 80, 153, 65, 254, 150, 242, 52, 54, 58, 218, 208, 217, 23, 137, 19, 160, 155, 171, 249, 72, 189)
	temp = append(temp, 202, 248, 54, 51, 14, 152, 217, 232, 149, 148, 42, 176, 146, 66, 1, 71, 224, 83, 167, 253, 101, 56, 106, 101, 107, 57, 60, 108, 202, 45, 81, 50, 51, 205, 104, 92, 223, 35, 174, 88)
	temp = append(temp, 79, 14, 125, 187, 219, 152, 86, 98, 196, 97, 148, 42, 221, 81, 205, 121, 235, 56, 153, 74, 74, 107, 166, 45, 123, 197, 186, 187, 146, 154, 244, 125, 183, 220, 87, 198, 253, 233, 102, 42)
	temp = append(temp, 143, 248, 4, 243, 40, 51, 12, 181, 18, 139, 57, 208, 170, 206, 174, 172, 144, 119, 45, 227, 242, 176, 35, 33, 167, 79, 38, 251, 81, 69, 158, 39, 195, 176, 106, 106, 11, 211, 253, 60)
	temp = append(temp, 49, 101, 5, 250, 33, 100, 160, 54, 30, 93, 139, 212, 233, 215, 225, 89, 114, 47, 12, 28, 71, 62, 28, 191, 224, 21, 192, 87, 162, 48, 71, 198, 77, 2, 189, 191, 111, 182, 252, 18)
	temp = append(temp, 16, 76, 59, 222, 234, 221, 123, 0, 0, 0, 0, 0)

	temp = append(temp, make([]byte, 512-len(temp))...)
	return temp
}
